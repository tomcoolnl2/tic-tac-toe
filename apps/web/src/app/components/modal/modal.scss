
$animation-speed: .7s;

@keyframes fade-in {
    0% { opacity: 0; }
    100% { opacity: 1; }
}

@keyframes slide-down {
    0% { top: -5%; }
    100% {  top: 0%; }
}

.modal-backdrop {
    animation: fade-in $animation-speed;
    position: absolute;
    inset: 0;
    background-color: rgba(black, .4);
}

.modal-background {
    animation: slide-down $animation-speed;
    background-color: var(--emerald-forest);
    height: 200px;
    width: 100%;
}

.modal-content {
    padding: 20px 0;
    width: 360px;
    height: 100%;
}
