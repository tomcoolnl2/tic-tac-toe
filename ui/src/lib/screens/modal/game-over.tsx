import React from 'react';
import useSound from 'use-sound';
import { type PlayerSymbol, GameStatus } from '@tic-tac-toe/model';
import type { AppScreenContent } from '../../context/content/model';
import { Divider, Grid } from '../../core';
import { GameOver, Button } from '../../components';

import drawSfx from '../../sound/draw.wav';
import looseSfx from '../../sound/loose.wav';
import winSfx from '../../sound/win.wav';

interface Props {
	content: AppScreenContent;
	gameStatus: GameStatus;
	playerSymbol: PlayerSymbol;
	cpuSymbol: PlayerSymbol;
	handleQuitGame: () => void;
	handleNextRound: () => void;
}

export const GameOverModalScreen: React.FC<Props> = ({
	content,
	playerSymbol,
	cpuSymbol,
	gameStatus,
	handleQuitGame,
	handleNextRound,
}) => {
	//
	const [playDrawSfx] = useSound(drawSfx, { volume: 0.75 });
	const [playLooseSfx] = useSound(looseSfx, { volume: 0.75 });
	const [playWinSfx] = useSound(winSfx, { volume: 0.5 });

	let title = content.title[0];
	let subtitle = '';
	let avatar: PlayerSymbol | null = null;
	let className = '';

	switch (gameStatus) {
		case GameStatus.WIN:
			subtitle = content.subtitle?.[0] || '';
			avatar = playerSymbol;
			className = ['x', 'o'][playerSymbol];
			playWinSfx();
			break;
		case GameStatus.LOST:
			subtitle = content.subtitle?.[1] || '';
			avatar = cpuSymbol;
			className = ['x', 'o'][cpuSymbol];
			playLooseSfx();
			break;
		case GameStatus.DRAW:
			title = content.title[1];
			subtitle = content.subtitle?.[2] || '';
			className = 'text-light';
			playDrawSfx();
			break;
	}

	return (
		<>
			<GameOver title={title} subtitle={subtitle} avatar={avatar} className={className} />
			<Divider margin="vertical-l" />
			<Grid cols={2} colGap="l">
				<Button variant="light" onClick={handleQuitGame}>
					{content.cta1}
				</Button>
				<Button variant="secondary" onClick={handleNextRound}>
					{content.cta2}
				</Button>
			</Grid>
		</>
	);
};
