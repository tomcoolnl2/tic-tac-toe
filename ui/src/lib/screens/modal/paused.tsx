import React from 'react';
import { useSaveGameManager } from '@tic-tac-toe/database';
import type { AppScreenContent } from '../../context/content/model';
import { Divider, Grid } from '../../core';
import { Button } from '../../components';
import { useInterfaceHandlers } from '../../hooks';

interface Props {
	content: AppScreenContent;
}

export const PausedModalScreen: React.FC<Props> = ({ content }) => {
	const { handleResumeGame } = useInterfaceHandlers();
	const { addSaveGame } = useSaveGameManager();
	return (
		<>
			<h1>{content.title}</h1>
			<Divider margin="vertical-l" />
			<Grid colGap="l">
				<Button variant="primary" onClick={handleResumeGame}>
					{content.cta2}
				</Button>
				<Button variant="secondary" onClick={handleResumeGame}>
					{content.cta1}
				</Button>
			</Grid>
		</>
	);
};
