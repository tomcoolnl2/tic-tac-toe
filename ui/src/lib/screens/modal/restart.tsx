import React from 'react';
import type { AppScreenContent } from '../../../../../content/src/lib/model';
import { Divider, Grid } from '../../core';
import { Button } from '../../components';

interface Props {
	content: AppScreenContent;
	handleQuitGame: () => void;
	handleResumeGame: () => void;
}

export const RestartModalScreen: React.FC<Props> = ({ content, handleQuitGame, handleResumeGame }) => {
	return (
		<>
			<h1>{content.title}</h1>
			<Divider margin="vertical-l" />
			<Grid cols={2} colGap="l">
				<Button variant="light" onClick={handleResumeGame}>
					{content.cta1}
				</Button>
				<Button variant="secondary" onClick={handleQuitGame}>
					{content.cta2}
				</Button>
			</Grid>
		</>
	);
};
