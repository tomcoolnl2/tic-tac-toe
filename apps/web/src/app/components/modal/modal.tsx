import React from 'react';
import { FlexBox } from '@tic-tac-toe/ui';
import './modal.scss';

export interface Props {
	children: React.ReactNode;
}

export const Modal: React.FC<Props> = ({ children }) => {
	return (
		<FlexBox className="modal-backdrop" direction="row" alignItems="center">
			<FlexBox
				className="modal-background"
				direction="column"
				alignItems="center"
			>
				<FlexBox
					className="modal-content"
					direction="column"
					alignItems="center"
					justifyContent="center"
				>
					{children}
				</FlexBox>
			</FlexBox>
		</FlexBox>
	);
};
