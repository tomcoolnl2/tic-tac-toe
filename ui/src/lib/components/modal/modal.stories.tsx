import type { StoryFn, Meta } from '@storybook/react';
import { type Props, Modal } from './modal';

const Story: Meta<typeof Modal> = {
	component: Modal,
	title: 'Components/Modal',
};
export default Story;

const Template: StoryFn<Props> = (args) => (
	<Modal {...args}>
		<h2>Modal header</h2>
		<br />
		<small>With some subtext</small>
	</Modal>
);

export const Default = Template.bind({});
Default.args = {};
