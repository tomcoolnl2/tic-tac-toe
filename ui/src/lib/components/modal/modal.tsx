import React from 'react';
import { FlexBox } from '../../core';
import './modal.scss';

export interface Props {
	children: React.ReactNode;
}

export const Modal: React.FC<Props> = React.memo(({ children }) => {
	return (
		<FlexBox className="modal-backdrop" direction="row" alignItems="center">
			<FlexBox className="modal-background" direction="column" alignItems="center">
				<FlexBox
					className="modal-content"
					direction="column"
					alignItems="center"
					justifyContent="center"
				>
					{children}
				</FlexBox>
			</FlexBox>
		</FlexBox>
	);
});
