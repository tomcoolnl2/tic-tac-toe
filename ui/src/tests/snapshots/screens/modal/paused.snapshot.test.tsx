import { render } from '@testing-library/react';
import { PausedModalScreen } from '../../../../lib/screens/modal/paused';

describe('PausedModalScreen component', () => {
	const content = {
		cta1: 'Resume',
		title: 'Game is Paused',
	};

	it('matches snapshot', () => {
		const { container } = render(<PausedModalScreen content={content} handleResumeGame={jest.fn()} />);
		expect(container.firstChild).toMatchSnapshot();
	});
});
