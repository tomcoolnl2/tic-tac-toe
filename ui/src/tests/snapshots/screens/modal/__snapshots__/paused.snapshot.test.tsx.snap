// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PausedModalScreen component matches snapshot 1`] = `
<h1>
  Game is Paused
</h1>
`;
